core/drivers/crypto/caam/caam_rng.c

```diff
@@ -0,0 +1,568 @@
+					   JUMP_C1_LOCAL(ALL_COND_TRUE,
+							 JMP_COND(NONE), 1));
+			caam_desc_add_word(desc,
+					   LD_NOCLASS_IMM(REG_CLEAR_WRITTEN,
+							  sizeof(uint32_t)));
+			caam_desc_add_word(desc, 0x1);
+			caam_desc_add_word(desc, RNG_SH_INST(sh_idx));
+		}
+	}
+
+	/* Load the Key if needed */
+	if (!key_loaded) {
+		/*
+		 * Add a wait loop while previous operation not completed,
+		 * followed by a register clear before executing next command
+		 */
+		caam_desc_add_word(desc, JUMP_C1_LOCAL(ALL_COND_TRUE,
+						       JMP_COND(NONE), 1));
+		caam_desc_add_word(desc, LD_NOCLASS_IMM(REG_CLEAR_WRITTEN,
+							sizeof(uint32_t)));
+		caam_desc_add_word(desc, 0x1);
+		caam_desc_add_word(desc, RNG_GEN_SECKEYS);
+	}
+
+	RNG_DUMPDESC(desc);
+}
+
+enum caam_status caam_rng_instantiation(void)
+{
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct caam_jobctx jobctx = {};
+	uint32_t *desc = NULL;
+	uint32_t sh_status = 0;
+	uint32_t nb_sh = 0;
+	uint32_t sh_mask = 0;
+	uint32_t inc_delay = 0;
+
+	RNG_TRACE("RNG Instantation");
+
+	/* Check if RNG is already instantiated */
+	if (caam_hal_rng_instantiated(rng_privdata->baseaddr)) {
+		RNG_TRACE("RNG already instantiated");
+		retstatus = CAAM_NO_ERROR;
+		goto end_inst;
+	}
+
+	/*
+	 * RNG needs to be instantiated. Allocate and prepare the
+	 * Job Descriptor
+	 */
+
+	/* Calculate the State Handles bit mask */
+	nb_sh = caam_hal_rng_get_nb_sh(rng_privdata->baseaddr);
+	sh_mask = GENMASK_32(nb_sh - 1, 0);
+
+	/*
+	 * The maximum size of the descriptor is:
+	 *    |----------------------|
+	 *    | Header               | = 1
+	 *    |----------------------|
+	 *    | First instantation   | = 1
+	 *    |----------------------|-------------------------
+	 *    | wait complete        | = 1
+	 *    |----------------------|
+	 *    | Clear done status    |       Repeat (nb_sh - 1)
+	 *    |                      | = 2
+	 *    |----------------------|
+	 *    | next SH instantation | = 1
+	 *    |----------------------|-------------------------
+	 *    | wait complete        | = 1
+	 *    |----------------------|
+	 *    | Clear done status    | = 2
+	 *    |                      |
+	 *    |----------------------|
+	 *    | Generate Secure Keys | = 1
+	 *    |----------------------|
+	 *    | Pad with a 0         | = 1
+	 */
+	desc = caam_calloc_desc(2 + (nb_sh - 1) * 4 + 4 + 1);
+	if (!desc) {
+		RNG_TRACE("Descriptor Allocation error");
+		retstatus = CAAM_OUT_MEMORY;
+		goto end_inst;
+	}
+
+	jobctx.desc = desc;
+
+	do {
+		/* Check if all State Handles are instantiated */
+		sh_status = caam_hal_rng_get_sh_status(rng_privdata->baseaddr);
+		if ((sh_status & sh_mask) == sh_mask) {
+			RNG_TRACE("RNG All SH are instantiated (0x%08" PRIx32
+				  ")",
+				  sh_status);
+			retstatus = CAAM_NO_ERROR;
+			goto end_inst;
+		}
+
+		if (sh_status == 0) {
+			retstatus = caam_hal_rng_kick(rng_privdata->baseaddr,
+						      inc_delay);
+			RNG_TRACE("RNG Kick (inc=%u) ret 0x%08" PRIx32,
```

Comment:
s/%u/%"PRIu32"/

s/`%08" PRIx32`/`%08x`/  (`retstatus` is an enum and will evaluate as an `int` or `unsigned int`, not explicitly a 32bit value.

Reply:
ok